File: PRACTICA_JAVA/src/irrgarten/Weapon.java
package irrgarten;

/**
 * Represents a weapon in the Irrgarten game.
 * Weapons have attack power and a number of uses before they are discarded.
 * 
 * @author marcosbslinux
 * @version 1.0
 */
public class Weapon {
    /** Attack power of the weapon */
    private float power;
    
    /** Number of remaining uses */
    private int uses;

    /**
     * Constructs a new Weapon with specified power and uses
     * @param p Attack power
     * @param u Number of uses
     */
    public Weapon(float p, int u) {
        power = p;
        uses = u;
    }

    /**
     * Uses the weapon to attack
     * @return Attack power if weapon has uses remaining, 0 otherwise
     */
    public float attack() {
        float damage = 0;
        if (uses > 0) {
            --uses;
            damage = power;
        }
        return damage;
    }

    /**
     * Generates string representation of weapon's state
     * @return String showing power and remaining uses
     */
    @Override
    public String toString() {
        return "W[" + Float.toString(power) + ", " + Integer.toString(uses) + "]";
    }

    /**
     * Determines if weapon should be discarded based on remaining uses
     * @return true if weapon should be discarded, false otherwise
     */
    public boolean discard() {
        return Dice.discardElement(uses);
    }
}

File: PRACTICA_RUBY/Weapon.rb
#encoding: utf-8
module Irrgarten
  # Represents a weapon in the Irrgarten game
  # Manages weapon power and remaining uses
  class Weapon
    # Initializes a new weapon with specified power and uses
    # @param p [Integer] attack power of the weapon
    # @param u [Integer] number of remaining uses
    def initialize(p, u)
      @power=p
      @uses=u
    end

    # Calculates attack damage from the weapon
    # Decreases uses by 1 if weapon has remaining uses
    # @return [Integer] attack power if weapon has uses, 0 otherwise
    def attack
      damage=0
      if @uses>0
        @uses -= 1
        damage=@power
      end
      damage
    end

    # Returns string representation of weapon state
    # @return [String] formatted weapon description showing power and uses
    def to_s
      "W[#{@power}, #{@uses}]"
    end

    # Determines if weapon should be discarded
    # Uses dice roll to decide based on remaining uses
    # @return [Boolean] true if weapon should be discarded
    def discard
      Dice.discard_element(@uses)
    end
  end
end
